doc/i3-match.1
.TH I3\-MATCH 1
.SH NAME
i3\-match \- access to i3 window manager's IPC interface
.SH SYNOPSIS
.B i3\-match
.RB [ \-S ]
.RI [ option | filter ]...
.RB [ \-o " \fIfield" \fR...]
.SH DESCRIPTION
.B i3\-match
provides command\-line access to i3 window manager's IPC (interprocess
communication) interface.
.PP
.B i3\-match
has two modes of operation: match\-mode and subscribe\-mode.
.PP
In both modes, filters are used to select nodes.
.PP
Match\-mode is the default. In this mode nodes are containers/windows in
i3's layout tree.
.PP
The option
.B \-S
enters subscribe\-mode. Here, nodes are events received via
i3's subscribe feature.
.SH OPTIONS
.TP
.B \-a
Print unmatched nodes too. Output the
.B :match
field or specify
.B \-h
to highlight matches.
If specified twice in subscribe mode ("almost all"), events not matching
filters for the
.B :evtype
field are left out.
.TP
.BI \-d " delimiter"
Set the delimiter. Default is " " (space).
.TP
.BI \-e " field op value"
Specify a filter with three following arguments. See
.B FILTERS
below.
.TP
.B \-h
Highlight matches. Useful with
.BR \-a .
.TP
.BI \-i " file"
Read json tree from
.I file
instead of querying via i3 IPC. If
.I file
is
.B \-
read from stdin. Match\-mode only.
.TP
.BI \-l " num"
Consider success after
.I num
matches. Default is 1. Match\-mode only (see also
.BR \-n ).
.TP
.B \-m
Keep monitoring for events infinitely. Specifying
.B \-n
after this option overrides this effect. Sets printed fields to
.B ":evtype change current/name container/name binding/command"
unless
.B \-o
is given.
Subscribe\-mode only.
.TP
.BI \-n " num"
In match\-mode: Exit after printing
.I num
nodes. If no output fields are selected, this option is ignored (you may want
.B \-l
instead).
In subscribe\-mode: exit after matching
.I num
nodes.
.TP
.BI \-o " field" \fR...
Print the values of given fields for each matched node. Printed values are
separated by the delimiter
.RB "(see " \-d ).
All arguments following this option are treated as field names.
.TP
.B \-S
Enter subscribe\-mode. This may not be specified after mode\-specific options.
.TP
.BI \-s " path"
Specify the path of the i3 socket to use. By default, uses the socket of the
i3 instance of the current display.
.TP
.B \-t
Same as
.BR "\-o :itree name" .
If both
.B \-t
and
.B \-o
are given,
.B :itree
is added before the first specified ouptut field. Match\-mode only.
.SH FILTERS
Filters are conditions applied to each node to choose which nodes to select.
All filters need to match to a node for it to be selected.
.PP
Filters are specified with the syntax
.I field op value
in a single argument without delimiting characters or by specifying the arguments
.B \-e
.IR "field op value" .
.I field
is the name of any field (see
.B FIELDS
below).
.I value
is the string to match against.
.I op
is any of the following operators:
.TP
.B =
Matches when the value of
.I field
is exactly
.IR value .
.TP
.B *=
Matches when the value of
.I field
contains
.IR value .
.TP
.B ^=
Matches when the value of
.I field
starts with
.IR value .
.TP
.B $=
Matches when the value of
.I field
ends with
.IR value .
.PP
The meaning of any operator can be inverted by prefixing it with
.B !
(exclamation mark).
.SH FIELDS
Fields contain values for each node. They are accessed via their names in
the json\-structure of each node.
.PP
Hierarchies of json objects are accessed using
.B /
(slashes) as separator.
.PP
Fields not found for a node are empty. There is no check for whether a field
name is valid.
.PP
See i3 IPC documentation (https://i3wm.org/docs/ipc.html) for a list of fields.
.SS "Pseudo fields"
.B i3\-match
provides additional information for each node. Pseudo\-fields access this
information. All pseudo\-fields start with colon
.B :
(colon).
.TP
.B :match
1, if this node is a match. 0 otherwise. 0 when used in a filter.
.TP
.B :level
Nesting level of this node in i3's layout tree. 0 for the root node.
0 in subscribe\-mode.
.TP
.B :floating
1, if this node is contained in a floating_nodes array. Increments by 1 for
each nesting level inside such a node. 0 in subscribe\-mode.
.TP
.B :scratch
1, if this node's scratchpad_state is not none. Increments by 1 for each
nesting level inside such a node. 0 in subscribe\-mode.
.TP
.B :workspace
Name of the workspace containing this node. Empty if this node is not
contained by a workspace. Empty in subscribe\-mode.
.TP
.B :output
Name of the output (screen) containing this node. Empty if this node
is not contained by an output. Empty in subscribe\-mode.
.TP
.B :sibi
Index of this node within the current list of siblings.
0 in subscribe\-mode.
.TP
.B :sibc
Number of siblings of this node. 0 in subscribe\-mode.
.TP
.B :childc
Number of children of this node. 0 in subscribe\-mode.
.TP
.B :indent
Outputs an indented hierarchy tree. Matched nodes are indented with '\-\-' (two
dashes) other nodes with '  ' (two spaces) per level. Only useful in
match\-mode.
.TP
.B :itree
Outputs an indented hierarchy resembling the output of
.BR tree (1).
Only useful in match\-mode.
.TP
.B :evtype
Event type in subscribe\-mode. One of
.BR workspace ", " output ", " mode ", " window ", " barconfig_update ", " binding .
.sp
This field is special when used in filters in subscribe\-mode. These filters
are used to decide which events to subscribe for. This an optimization and
only makes a visual difference when specifying
.B \-a
twice.
.sp
When filters are specified such that no event types would match,
.B i3\-match
exits with an error message.
.sp
"none" in match\-mode.
.TP
.B :nodei
Index of this node.
.TP
.B :matchc
Number of matches up to this point. This is incremented after
a node is matched and before processing output.
.TP
.B :json
Json representation of this node formatted without newlines.
.TP
.BI :json: field
Json representation of
.I field
formatted without newlines. See
.B FIELDS
above for
.IR field .
Pseudo fields are not available here.
.SH "ENVIRONMENT VARIABLES"
.TP
.B I3SOCK
If
.B \-s
is not given, this variable is used to determine the path of i3's IPC socket,
overriding the usual lookup.
.SH "EXIT STATUS"
The exit status is 0 on success, 1 if success condition was not met and
2 if an error occurred.
In match\-mode success means that at least one node was selected. If
.B \-l
was specified, the exit status is according to that option.
In subscribe\-mode
.B i3\-match
exits on success, so the exit status is always 0 unless an error occurred.
.SH EXAMPLES
.TP
.B i3\-match \-t
Print a simple hierarchy tree with names for each container.
.TP
.B i3\-match 'name*=Firefox' \-l 2 && echo true
Print "true" when there are at least two containers with titles
containing "Firefox".
.TP
.B i3\-match focused=true \-a \-h \-o :itree rect/{width,height} name
Print width, height and title for each window and container.
Highlight the focused window.
.TP
.B i3\-match \-S :evtype=window change=new
Wait for a window to be created, then exit.
.TP
.B i3\-match \-Sm :evtype=window container/focused=true \-o container/name
Print the name of the focused window every time it changes.


src/i3-match.c
/*
 * A program for matching containers in the i3 window manager.
 *
 * needed libraries: yajl
 *
 * Created:     2014-03-11
 * Last Change: 2014-03-14
 *
 */

#include "i3util.h"
#include "i3ipc.h"
#include "i3json.h"
#include "debug.h"
#include "jsonutil.h"
#include "util.h"
#include "sb.h"

#include <assert.h>
#include <unistd.h>
#include <stdlib.h>
#include <string.h>

#ifndef ASCII_TREE
// Unicode version of indent tree strings.
// Strings used for nodes.
#define IT_MID_CHILD "\u251c\u2500"
#define IT_LAST_CHILD "\u2514\u2500"
// String used for the root node. Should be one
// cell shorter than IT_BAR and IT_EMPTY.
#define IT_ROOT "\u2500"
// Strings used for indenting, the last character
// is removed when indenting for the root node.
#define IT_BAR "\u2502  "
#define IT_EMPTY "   "
#else
// Ascii version of indent tree strings.
#define IT_ROOT "-"
#define IT_MID_CHILD "+-"
#define IT_BAR "|  "
#define IT_EMPTY "   "
#define IT_LAST_CHILD "`-"
#endif

#define COL_BOLD "\x1b[1m"
#define COL_NORMAL "\x1b[m"

#define CH_SPACE ' '
#define CH_BAR '|'

const char const *ALL_EVENTS_SUB_JSON = "[\"workspace\",\"output\",\"mode\",\"window\",\"barconfig_update\",\"binding\"]";

#define EVENT_TYPE_COUNT 6
const char const *EVENT_NAMES[EVENT_TYPE_COUNT] = {
    "workspace", // (I3_IPC_EVENT_MASK | 0)
    "output", // (I3_IPC_EVENT_MASK | 1)
    "mode", // (I3_IPC_EVENT_MASK | 2)
    "window", // (I3_IPC_EVENT_MASK | 3)
    "barconfig_update", // (I3_IPC_EVENT_MASK | 4)
    "binding", // (I3_IPC_EVENT_MASK | 5)
};

#define TREE_OUTPUTS_COUNT 2
char *TREE_OUTPUTS[TREE_OUTPUTS_COUNT] = { ":itree", "name" };

#define DEFAULT_MONITOR_OUTPUT_COUNT 5
char *DEFAULT_MONITOR_OUTPUTS[DEFAULT_MONITOR_OUTPUT_COUNT] = {
    ":evtype", "change", "current/name", "container/name", "binding/command"
};

typedef enum mode {
    MODE_MATCH = 0,
    MODE_SUBSCRIBE = 1,
} mode;

typedef enum outmode {
    OUT_NONE = 0,
    OUT_FIELDS = 1,
} outmode;

enum flags {
    F_PRINTALL = 1<<0,
    F_HIGHLIGHT = 1<<1,
};

typedef struct context {
    i3json_matcher *matchers;
    int matcherc;
    outmode outmode;
    char **outputs;
    int outputc;
    int flags;
    int maxcount;
    char *field_sep;
    size_t field_sep_len;
    string_builder sb;
    string_builder itree;
    int objcount;
    int matchcount;
    i3json_print_tree_context pt_context;
    i3_msg *msg;
    // yajl generator on sb.
    yajl_gen sb_gen;
} context;

static const char *get_eventtype(unsigned int type) {
    unsigned int i = type & ~I3_IPC_EVENT_MASK;
    if (i != type) {
        if (i >= EVENT_TYPE_COUNT) {
            return "unknown";
        }
        return EVENT_NAMES[i];
    }
    return "none";
}

static yajl_gen get_sb_gen(context *context) {
    yajl_gen gen = context->sb_gen;
    if (!gen) {
        gen = yajl_gen_alloc(NULL);
        int res = yajl_gen_config(gen, yajl_gen_print_callback,
                                  yajlutil_print_cb_sb_push, &context->sb);
        assert(res);
        context->sb_gen = gen;
    } else {
        yajl_gen_reset(gen, NULL);
    }
    return gen;
}

static void push_value(string_builder *sb, const char* key,
                       yajl_val node, iter_info *info,
                       context *ctx, int match) {
    debug_print("key=%s\n", key);
    if (strcmp(":match", key) == 0) {
        sb_pushf(sb, "%d", !!match);
    } else if (strcmp(":level", key) == 0) {
        sb_pushf(sb, "%d", info->level);
    } else if (strcmp(":floating", key) == 0) {
        sb_pushf(sb, "%d", ctx->pt_context.floating);
    } else if (strcmp(":scratch", key) == 0) {
        sb_pushf(sb, "%d", ctx->pt_context.scratch);
    } else if (strcmp(":workspace", key) == 0) {
        if (ctx->pt_context.ws) {
            sb_pushf(sb, "%s", ctx->pt_context.ws);
        }
    } else if (strcmp(":output", key) == 0) {
        if (ctx->pt_context.output) {
            sb_pushf(sb, "%s", ctx->pt_context.output);
        }
    } else if (strcmp(":sibi", key) == 0) {
        sb_pushf(sb, "%d", info->nodei);
    } else if (strcmp(":sibc", key) == 0) {
        sb_pushf(sb, "%d", info->nodec);
    } else if (strcmp(":childc", key) == 0) {
        sb_pushf(sb, "%d", info->subnodec);
    } else if (strcmp(":indent", key) == 0) {
        const char *str = match ? "--" : "  ";
        int j = info->level; // indent level + 1 times
        for (; j >= 0; j--) {
            sb_pushn(sb, str, 2);
        }
    } else if (strcmp(":itree", key) == 0) {
        string_builder *isb = &ctx->itree;
        int n = info->level;
        int j;
        for (j = 0; j < n; j++) {
            if (isb->buf[j] == CH_SPACE) {
                sb_pushn(sb, IT_EMPTY, strlen(IT_EMPTY) - !j);
            } else {
                sb_pushn(sb, IT_BAR, strlen(IT_BAR) - !j);
            }
        }
        if (info->level == 0) {
            sb_push(sb, IT_ROOT);
        } else if (info->nodei < info->nodec - 1) {
            sb_push(sb, IT_MID_CHILD);
        } else {
            sb_push(sb, IT_LAST_CHILD);
        }
    } else if (strcmp(":evtype", key) == 0) {
        i3_msg *msg = ctx->msg;
        if (msg) {
            sb_push(sb, get_eventtype(msg->type));
        } else {
            sb_pushn(sb, "none", 4);
        }
    } else if (strcmp(":nodei", key) == 0) {
        sb_pushf(sb, "%d", ctx->objcount);
    } else if (strcmp(":matchc", key) == 0) {
        sb_pushf(sb, "%d", ctx->matchcount);
    } else if (strncmp(":json", key, 5) == 0 && (key[5] == '\0' || key[5] == ':')) {
        yajl_val n = node;
        if (key[5]) {
            n = yajlutil_path_get(node, key + 6, yajl_t_any);
        }
        if (n) {
            yajlutil_serialize_val(get_sb_gen(ctx), n, 0);
        } else {
            sb_pushn(sb, "null", 4);
        }
    } else {
        yajl_val n = yajlutil_path_get(node, key, yajl_t_any);
        const char *str = yajlutil_get_string(n);
        if (str) {
            sb_push(sb, str);
        } else {
            // n is array or hash
            yajlutil_serialize_val(get_sb_gen(ctx), n, 0);
        }
    }
}

static void format_fields(yajl_val node, iter_info *info, context *ctx, int match) {
    int i;
    string_builder *sb = &ctx->sb;
    if (match && ctx->flags & F_HIGHLIGHT) {
        sb_push(sb, COL_BOLD);
    }
    for (i = 0; i < ctx->outputc; i++) {
        if (i) sb_pushn(sb, ctx->field_sep, ctx->field_sep_len);
        const char *key = ctx->outputs[i];
        push_value(sb, key, node, info, ctx, match);
    }
    if (match && ctx->flags & F_HIGHLIGHT) {
        sb_push(sb, COL_NORMAL);
    }
}

static void accum_itree(iter_info *info, context *context) {
    sb_trunc(&context->itree, info->level);
    char ch = CH_SPACE;
    if (info->level && info->nodei < info->nodec - 1) {
        ch = CH_BAR;
    }
    sb_pushn(&context->itree, &ch, 1);
}

typedef struct node_getter_args {
    yajl_val node;
    iter_info *info;
    context *ctx;
} node_getter_args;

static const char *node_value_getter(const char *key, void *ptr) {
    node_getter_args *args = ptr;
    string_builder *sb = &args->ctx->sb;
    sb_trunc(sb, 0);
    push_value(sb, key, args->node, args->info, args->ctx, 1);
    return sb->buf;
}

static iter_advise process_node(yajl_val node, iter_info *info, context *ctx) {
    ++ctx->objcount;
    node_getter_args gargs = {
        .node = node,
        .info = info,
        .ctx = ctx,
    };
    int match = 0;
    if (i3json_matchers_match_ex(ctx->matcherc, ctx->matchers, &node_value_getter, &gargs)) {
        ++ctx->matchcount;
        match = 1;
    }
    accum_itree(info, ctx);
    if (match || ctx->flags & F_PRINTALL) {
        switch (ctx->outmode) {
        case OUT_NONE:
            break;
        case OUT_FIELDS:
            sb_trunc(&ctx->sb, 0);
            format_fields(node, info, ctx, match);
            sb_pushn(&ctx->sb, "\n", 1);
            fwrite(ctx->sb.buf, sizeof(char), ctx->sb.len, stdout);
            break;
        }
    }
    if (match && ctx->maxcount > 0 && ctx->matchcount >= ctx->maxcount) {
        return ITER_ABORT_SUCCESS;
    }
    return ITER_CONT;
}

static iter_advise iter_pred(yajl_val node, iter_info *info, void *ptr) {
    context *ctx = ptr;
    i3json_tree_accum_data(node, info, &ctx->pt_context);
    return process_node(node, info, ctx);
}

static int eventloop(int sock, context *ctx) {
    i3_msg msg = EMPTY_I3_MSG;
    yajl_val event = NULL;
    char errbuf[ERROR_BUFSIZ];
    // iter_info values not meaningful for events
    iter_info info = { 0, 0, 0, 0, 0 };
    ctx->msg = &msg;
    for (;;) {
        if (i3ipc_recv_message(sock, &msg) == -1) {
            del_i3_msg(&msg);
            return 2;
        }
        event = yajl_tree_parse(msg.data, errbuf, sizeof(errbuf));
        if (!event) {
            print_error("parse error", errbuf);
        }
        iter_advise advise = process_node(event, &info, ctx);
        yajl_tree_free(event);
        i3ipc_msg_recycle(&msg);
        event = NULL;
        switch (advise) {
        case ITER_ABORT_SUCCESS:
            return 0;
        case ITER_ABORT:
            return 1;
        case ITER_CONT:
            break;
        case ITER_NODESC:
            assert(0);
        }
    }
    del_i3_msg(&msg);
    ctx->msg = NULL;
}

static int match_evtype_only(i3json_matcher *matchers, int matcherc, const char *type) {
    int i;
    for (i = 0; i < matcherc; i++) {
        i3json_matcher *matcher = matchers + i;
        if (i3json_matcher_cmp_key(matcher, ":evtype") == 0) {
            if (!i3json_matcher_match_value(type, matcher)) {
                return 0;
            }
        }
    }
    return 1;
}

static char *get_matching_evtypes(i3json_matcher *matchers, int matcherc) {
    string_builder sb = EMPTY_STRING_BUILDER;
    yajl_gen gen = yajl_gen_alloc(NULL);
    int res = yajl_gen_config(gen, yajl_gen_print_callback,
                              yajlutil_print_cb_sb_push, &sb);
    assert(res);
    int i;
    int matches = 0;
    yajl_gen_array_open(gen);
    for (i = 0; i < EVENT_TYPE_COUNT; i++) {
        if (match_evtype_only(matchers, matcherc, EVENT_NAMES[i])) {
            ++matches;
            debug_print("%s matches\n", EVENT_NAMES[i]);
            yajl_gen_string(gen,
                            (const unsigned char *) EVENT_NAMES[i],
                            strlen(EVENT_NAMES[i]));
        }
    }
    if (!matches) {
        sb_free(&sb);
        return NULL;
    }
    yajl_gen_array_close(gen);
    return sb_disown(&sb);
}

int main(int argc, char *argv[]) {
    i3json_matcher matchers[argc];

    #define SMALL_ITREE_SIZE 16
    char itree[SMALL_ITREE_SIZE];
    context context = {
        .matchers = matchers,
        .matcherc = 0,
        .outmode = OUT_NONE,
        .outputs = NULL,
        .outputc = 0,
        .flags = 0,
        .maxcount = 0,
        .field_sep = " ",
        .field_sep_len = 1,
        .sb = EMPTY_STRING_BUILDER,
        .itree = SB_WITH_EXT_BUF(itree, SMALL_ITREE_SIZE),
        .matchcount = 0,
        .pt_context = I3JSON_EMPTY_PRINT_TREE_CONTEXT,
        .sb_gen = NULL,
    };

    debug_print("BUFSIZ=%d\n", BUFSIZ);

    #define EXIT_MODE_ERROR(mode, option) \
        do { fprintf(stderr, option " can only be used in " mode "\n"); return 2; } while (0)
    const char *spath = NULL;
    mode mode = MODE_MATCH;
    int almostall = 0, mincount = 1, printtree = 0;
    char **aoutputs = NULL;
    int c;
    int have_modearg = 0;
    const char *infile = NULL;
    optind = 1;
    while (optind < argc) {
        int prevind = optind;
        if ((c = getopt(argc, argv, "+s:Si:ahml:n:d:e:to")) != -1) {
            debug_print("option: ind=%d c=%c\n", optind, c);
            switch (c) {
            case 's':
                spath = optarg;
                break;
            case 'S':
                if (have_modearg) {
                    fprintf(stderr, "cannot specify mode after mode-specific arguments\n");
                    return 2;
                }
                context.maxcount = 1;
                mode = MODE_SUBSCRIBE;
                break;
            case 'i':
                if (mode != MODE_MATCH) EXIT_MODE_ERROR("match-mode", "-i");
                have_modearg = 1;
                infile = optarg;
                break;
            case 'a':
                if (context.flags & F_PRINTALL) {
                    almostall = 1;
                }
                context.flags |= F_PRINTALL;
                break;
            case 'h':
                context.flags |= F_HIGHLIGHT;
                break;
            case 'm':
                if (mode != MODE_SUBSCRIBE) EXIT_MODE_ERROR("subscribe-mode", "-m");
                context.maxcount = -1;
                context.outmode = OUT_FIELDS;
                context.outputs = DEFAULT_MONITOR_OUTPUTS;
                context.outputc = DEFAULT_MONITOR_OUTPUT_COUNT;
                break;
            case 'l': {
                if (mode != MODE_MATCH) EXIT_MODE_ERROR("match-mode", "-l");
                int i, c;
                if (sscanf(optarg, "%d%n", &i, &c) < 1 || optarg[c] != 0) {
                    fprintf(stderr, "invalid min count (-l) - %s\n", optarg);
                    return 2;
                }
                have_modearg = 1;
                mincount = i;
                break;
            }
            case 'n': {
                int i, c;
                if (sscanf(optarg, "%d%n", &i, &c) < 1 || optarg[c] != 0) {
                    fprintf(stderr, "invalid max count (-n) - %s\n", optarg);
                    return 2;
                }
                have_modearg = 1;
                context.maxcount = i;
                break;
            }
            case 'd':
                context.field_sep = optarg;
                context.field_sep_len = strlen(optarg);
                break;
            case 'e': {
                have_modearg = 1;
                if (argc - optind < 2) {
                    fprintf(stderr, "missing %d arguments for -e\n", 2 + optind - argc);
                    return 2;
                }
                int op = i3json_parse_operator(argv[optind]);
                if (op == -1) {
                    fprintf(stderr, "invalid operator: %s\n", argv[optind]);
                    return 2;
                }
                i3json_make_matcher(optarg, argv[optind + 1], op, matchers + context.matcherc);
                ++context.matcherc;
                optind += 2;
                break;
            }
            case 't':
                if (mode != MODE_MATCH) EXIT_MODE_ERROR("match-mode", "-t");
                have_modearg = 1;
                printtree = 1;
                context.outmode = OUT_FIELDS;
                context.outputs = TREE_OUTPUTS;
                context.outputc = TREE_OUTPUTS_COUNT;
                break;
            case 'o':
                if (optind == prevind) {
                    fprintf(stderr, "no option allowed after -o\n");
                    return 2;
                }
                have_modearg = 1;
                context.outmode = OUT_FIELDS;
                if (printtree) {
                    aoutputs = calloc(argc - optind + 1, sizeof(char*));
                    aoutputs[0] = ":itree";
                    memcpy(aoutputs + 1, argv + optind, (argc - optind) * sizeof(char*));
                    context.outputs = aoutputs;
                    context.outputc = argc - optind + 1;
                } else {
                    context.outputs = argv + optind;
                    context.outputc = argc - optind;
                }
                goto argparse_finished;
            case '?':
                return 2;
            default:
                assert(0);
            }
        } else {
            have_modearg = 1;
            const char *arg = argv[optind];
            if (i3json_parse_matcher(arg, matchers + context.matcherc) == -1) {
                fprintf(stderr, "invalid filter: %s\n", arg);
                return 2;
            }
            ++context.matcherc;
            ++optind;
        }
    }
argparse_finished: {}

    int result;

    switch (mode) {
    case MODE_MATCH: {
        if (context.outmode == OUT_NONE) {
            context.maxcount = mincount;
        }
        string_builder buf = EMPTY_STRING_BUILDER;
        i3_msg msg = EMPTY_I3_MSG;
        yajl_val tree = NULL;
        if (infile) {
            FILE *stream = NULL;
            FILE *f = NULL;
            if (strcmp(infile, "-") == 0) {
                stream = stdin;
            } else {
                f = stream = fopen(infile, "r");
                if (!f) {
                    perror("open");
                    return 2;
                }
            }
            push_whole_file(&buf, stream);
            if (ferror(stream)) {
                sb_free(&buf);
                if (f) fclose(f);
                return 2;
            }
            if (buf.len == 0) {
                fprintf(stderr, "json input is empty\n");
                if (f) fclose(f);
                return 2;
            }
            char errbuf[ERROR_BUFSIZ];
            tree = yajl_tree_parse(buf.buf, errbuf, sizeof(errbuf));
            if (!tree) {
                print_error("json error", errbuf);
                sb_free(&buf);
                if (f) fclose(f);
                return 2;
            }
            if (f) fclose(f);
        } else {
            set_default_sigchld_handler();
            int sock = i3ipc_open_socket(spath);
            if (sock == -1) {
                return 2;
            }
            if (i3util_request_json(sock, I3_IPC_MESSAGE_TYPE_GET_TREE, "", &msg, &tree) == -1) {
                close(sock);
                return 2;
            }
            debug_print("%s\n", "close sock...");
            close(sock);
        }
        i3json_iter_nodes(tree, &iter_pred, &context);
        yajl_tree_free(tree);
        del_i3_msg(&msg);
        sb_free(&buf);
        result = context.matchcount >= mincount ? 0 : 1;
        break;
    }
    case MODE_SUBSCRIBE: {
        set_default_sigchld_handler();
        int sock = i3ipc_open_socket(spath);
        if (sock == -1) {
            return 2;
        }
        {
            const char *body = NULL;
            char *astr = NULL;
            if (context.flags & F_PRINTALL && !almostall) {
                body = ALL_EVENTS_SUB_JSON;
            } else {
                body = astr = get_matching_evtypes(context.matchers, context.matcherc);
                if (!body) {
                    fprintf(stderr, ":evtype never matches\n");
                    return 2;
                }
            }
            debug_print("body=%s\n", body);
            if (i3util_subscribe(sock, body) == -1) {
                fprintf(stderr, "subscribe request failed\n");
                free(astr);
                return 2;
            }
            free(astr);
        }
        result = eventloop(sock, &context);
        close(sock);
        break;
    }
    default:
        assert(0);
    }

    debug_print("%s\n", "cleanup...");
    if (context.sb_gen) {
        yajl_gen_free(context.sb_gen);
    }
    sb_free(&context.itree);
    sb_free(&context.sb);
    free(aoutputs);

    return result;
}
